feat: Add S3 chat memory repository implementation

ybezsonov opened the pull request:
Add S3ChatMemoryRepository with CRUD operations for chat history

**Features:**
- Implement S3ChatMemoryConfig with builder pattern and validation
- Add Spring Boot auto-configuration for integration
- Support configurable storage class, key prefix, and auto bucket initialization

**Test suite (15 integration tests):**
- Full CRUD operations, pagination, conversation replacement
- Metadata and message order preservation
- Unicode content, large messages (100KB), special characters handling
- Auto bucket initialization and error handling
- Storage class configuration

**Documentation:**
- README with usage examples and configuration options

Resolves: #5088

ybezsonov committed:
feat: Add S3 chat memory repository implementation

- Add S3ChatMemoryRepository with CRUD operations for chat history
- Implement S3ChatMemoryConfig with builder pattern and validation
- Add Spring Boot auto-configuration for integration
- Include test suite (22 tests total):
  - Property-based tests with jqwik for edge cases
  - Integration tests with Testcontainers and LocalStack
  - AutoConfiguration tests for Spring Boot integration
- Add documentation

Resolves: #5088
Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

ybezsonov committed:
refactor: Replace property-based tests with IT, improve S3ChatMemoryRepository

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

tzolov committed:
chore: Add jspecify null annotations to S3 chat memory module

- Add @Nullable annotations to Builder fields in S3ChatMemoryConfig
- Add @Nullable to extractConversationId return type in S3ChatMemoryRepository
- Add @NullMarked annotation to package-info.java
- Remove misplaced AutoConfiguration.imports file from core module

Signed-off-by: Christian Tzolov <christian.tzolov@broadcom.com>

ericbottard commented on memory/repository/spring-ai-model-chat-memory-repository-s3/src/test/java/org/springframework/ai/chat/memory/repository/s3/S3ChatMemoryRepositoryIT.java:
```diff
@@ -0,0 +1,378 @@
+ * @author Yuriy Bezsonov
+ */
+@Testcontainers
+class S3ChatMemoryRepositoryIT {
```
This class is a test, but is in src/main/java !!?

ybezsonov committed:
Merge branch 'spring-projects:main' into s3-memory

ybezsonov committed:
chore: Add jspecify null annotations to S3 chat memory autoconfigure module

- Add @NullMarked annotation to package-info.java
- Add @Nullable to getBucketName() getter (field without default)
- Remove duplicate maven-compiler-plugin configuration (inherited from parent)

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

ybezsonov committed:
chore: Add @Nullable to getBucketName() getter (field without default)

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

ybezsonov committed:
Merge branch 'spring-projects:main' into s3-memory

ybezsonov committed:
chore: Fix JSpecify type-use annotation style for null safety

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

ybezsonov committed:
Merge branch 'spring-projects:main' into s3-memory

ybezsonov committed:
Merge branch 'spring-projects:main' into s3-memory

ybezsonov committed:
docs: Add AWS SDK v2 requirements and findConversationIds() performance notes

- Document AWS SDK for Java 2.x requirement (v1 not supported)
- Add performance note about findConversationIds() pagination at scale
- Use ChatMemory.DEFAULT_CONVERSATION_ID for null/empty conversation IDs

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

akukuruza reviewed (approved):
Great job! Thank you!

akukuruza commented:
> > Hey @ybezsonov, I got one more follow-up question – is there any way to support dynamic key prefixes, or would that be too complex to add?
> > For example, if we wanted to organize chat history like `/memory/{product}/{date}/{conversationId}.json` where product and date are determined at runtime based on conversation metadata. Currently, the keyPrefix appears to be static configuration.
> > This would be helpful for multi-tenant scenarios or when we need to partition history by different business dimensions without creating separate repository instances for each combination.
> > Thanks!
> 
> Hi @akukuruza, This is an interesting idea, and multi-tenancy is a very valid use case. I don't know about the date, but I think sessionId could be useful in this case. Some time ago, I implemented short-term and long-term memory with Spring AI https://dev.to/yuriybezsonov/a-practical-guide-to-building-ai-agents-with-java-and-spring-ai-part-2-add-memory-odn and recently updated it to managed memory with AgentCore https://dev.to/yuriybezsonov/ai-agent-memory-made-easy-amazon-bedrock-agentcore-memory-with-spring-ai-3bng.
> 
> I will review them all and come up with a solution which should cover multi-tenancy and be ready for different types of memory - similar to namespaces in AgentCore, I think.

Appreciate your work on this! The way I see it, adding a `Function<String, String> keyResolver` to the config would enable users to encode tenant or runtime context into the conversation ID itself, letting the resolver transform it to the proper S3 namespace at runtime.

ybezsonov committed:
feat: Add dynamic key resolution for S3ChatMemoryRepository

Add optional keyResolver and conversationIdExtractor functions to
S3ChatMemoryConfig, enabling hierarchical S3 key structures at runtime.
This supports multi-tenant scenarios and business-dimension partitioning
without requiring separate repository instances per namespace.

When neither function is set, existing static keyPrefix behavior is
preserved. When set, both must be provided together (validated at build
time). The keyPrefix continues to serve as the listing base for
findConversationIds().

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

ybezsonov committed:
Merge remote-tracking branch 'upstream/main' into s3-memory

ybezsonov committed:
Fix copyright header to use 2023-present in S3 chat memory repository

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

ybezsonov committed:
Fix copyright header to use 2023-present in S3 autoconfigure module

Signed-off-by: Yuriy Bezsonov <ybezsonov@gmail.com>

sdeleuze commented:
As discussed with @tzolov, since Spring AI 2.1 plans to deprecate the memory abstraction in favor of a session one currently incubated in https://github.com/spring-ai-community/spring-ai-session and planned for inclusion in 2.1, we think this PR should be reworked to be integrated on session side to be merged in https://github.com/spring-ai-community/spring-ai-session initially and then in 2.1. As a consequence, I will close this PR unmerged. Thanks for your understanding, feel free to reach @tzolov or me for more related discussions.
